File: app/src/main/java/com/dmitrybrant/android/mandelbrot/MandelbrotNative.kt
package com.dmitrybrant.android.mandelbrot

import java.nio.ByteBuffer

class OrbitResult(
    val orbit: ByteBuffer,
    val polyScaled: FloatArray,
    val polyLim: Int,
    val polyScaleExp: Int,
    val radiusExp: Double
)

object MandelbrotNative {
    init {
        System.loadLibrary("mandel_native")
    }

    // Native methods
    external fun createState(x: Double, y: Double, r: Double, iterations: Int): Long
    external fun destroyState(statePtr: Long)
    external fun setState(statePtr: Long, x: Double, y: Double, r: Double, iterations: Int)
    external fun setStateStr(statePtr: Long, x: String, y: String, r: String, iterations: Int)
    external fun zoomIn(statePtr: Long, dx: Double, dy: Double, factor: Double)
    external fun zoomOut(statePtr: Long, factor: Double)
    external fun generateOrbit(statePtr: Long): OrbitResult?
    external fun setIterations(statePtr: Long, iterations: Int)
    external fun getCenterX(statePtr: Long): String?
    external fun getCenterY(statePtr: Long): String?
    external fun getRadius(statePtr: Long): String?

    const val ITERATIONS = 2000
    const val INIT_X = -0.5
    const val INIT_Y = 0.0
    const val INIT_R = 2.0

    class MandelbrotState {
        private var nativePtr: Long

        var iterations = ITERATIONS
            set(value) {
                field = value
                setIterations(nativePtr, iterations)
            }

        var cmapscale = 20.0

        init {
            nativePtr = createState(INIT_X, INIT_Y, INIT_R, iterations)
        }

        fun destroy() {
            if (nativePtr != 0L) {
                destroyState(nativePtr)
                nativePtr = 0
            }
        }

        @Throws(Throwable::class)
        protected fun finalize() {
            destroy()
        }

        fun set(x: Double, y: Double, r: Double, iterations: Int) {
            setState(nativePtr, x, y, r, iterations)
        }

        fun set(x: String, y: String, r: String, iterations: Int) {
            setStateStr(nativePtr, x, y, r, iterations)
        }

        fun zoomIn(dx: Double, dy: Double, factor: Double) {
            zoomIn(nativePtr, dx, dy, factor)
        }

        fun generateOrbit(): OrbitResult {
            return generateOrbit(nativePtr)!!
        }

        val centerX: String
            get() = getCenterX(nativePtr)!!

        val centerY: String
            get() = getCenterY(nativePtr)!!

        val radius: String
            get() = getRadius(nativePtr)!!

        fun reset() {
            iterations = ITERATIONS
            cmapscale = 20.0
            set(INIT_X, INIT_Y, INIT_R, iterations)
        }

        fun zoomOut(factor: Double) {
            zoomOut(nativePtr, factor)
        }
    }
}

File: ios/MandelbrotApp/AdvancedMode/MandelbrotNative.swift
import Foundation

struct OrbitResult {
    let orbitData: UnsafeMutablePointer<Float>
    let orbitSize: Int
    let polyScaled: [Float]
    let polyLim: Int
    let polyScaleExp: Int
    let radiusExp: Double
}

class MandelbrotState {
    static let ITERATIONS = 2048
    static let INIT_X = -0.5
    static let INIT_Y = 0.0
    static let INIT_R = 2.0
    static let INIT_COLOR_SCALE: Float = 23.0

    private var nativePtr: OpaquePointer?

    var numIterations: Int = ITERATIONS {
        didSet {
            numIterations = max(numIterations, 2)
            if let ptr = nativePtr {
                mandel_setIterations(UnsafeMutableRawPointer(ptr), Int32(numIterations))
            }
        }
    }

    init() {
        let ptr = mandel_createState(MandelbrotState.INIT_X, MandelbrotState.INIT_Y, MandelbrotState.INIT_R, Int32(numIterations))
        nativePtr = OpaquePointer(ptr)
    }

    deinit {
        destroy()
    }

    func destroy() {
        if let ptr = nativePtr {
            mandel_destroyState(UnsafeMutableRawPointer(ptr))
            nativePtr = nil
        }
    }

    func set(x: Double, y: Double, r: Double, iterations: Int) {
        guard let ptr = nativePtr else { return }
        mandel_setState(UnsafeMutableRawPointer(ptr), x, y, r, Int32(iterations))
        numIterations = iterations
    }

    func set(x: String, y: String, r: String, iterations: Int) {
        guard let ptr = nativePtr else { return }
        mandel_setStateStr(UnsafeMutableRawPointer(ptr), x, y, r, Int32(iterations))
        numIterations = iterations
    }

    func zoomIn(dx: Double, dy: Double, factor: Double) {
        guard let ptr = nativePtr else { return }
        mandel_zoomIn(UnsafeMutableRawPointer(ptr), dx, dy, factor)
    }

    func zoomOut(factor: Double) {
        guard let ptr = nativePtr else { return }
        mandel_zoomOut(UnsafeMutableRawPointer(ptr), factor)
    }

    func generateOrbit() -> OrbitResult {
        let result = mandel_generateOrbit(UnsafeMutableRawPointer(nativePtr!))
        let t = result.polyScaled
        let polyScaled: [Float] = [t.0, t.1, t.2, t.3, t.4, t.5]
        return OrbitResult(
            orbitData: result.orbitData,
            orbitSize: Int(result.orbitSize),
            polyScaled: polyScaled,
            polyLim: Int(result.polyLim),
            polyScaleExp: Int(result.polyScaleExp),
            radiusExp: result.radiusExp
        )
    }

    var centerX: String {
        guard let ptr = nativePtr else { return "" }
        return String(cString: mandel_getCenterX(UnsafeMutableRawPointer(ptr)))
    }

    var centerY: String {
        guard let ptr = nativePtr else { return "" }
        return String(cString: mandel_getCenterY(UnsafeMutableRawPointer(ptr)))
    }

    var radius: String {
        guard let ptr = nativePtr else { return "" }
        return String(cString: mandel_getRadius(UnsafeMutableRawPointer(ptr)))
    }

    func reset() {
        set(x: MandelbrotState.INIT_X, y: MandelbrotState.INIT_Y, r: MandelbrotState.INIT_R, iterations: MandelbrotState.ITERATIONS)
    }
}
